File: src/renderer/texture/GPUTextureView.cpp
////////////////////////////////////////////////////////////////////////////////
/// Reaper
///
/// Copyright (c) 2015-2022 Thibault Schueller
/// This file is distributed under the MIT License
////////////////////////////////////////////////////////////////////////////////

#include "GPUTextureView.h"

#include "core/Assert.h"

namespace Reaper
{
namespace
{
    u32 default_view_aspect(PixelFormat format)
    {
        switch (format)
        {
        case PixelFormat::D16_UNORM:
        case PixelFormat::D32_SFLOAT:
            return ViewAspect::Depth;
        case PixelFormat::X8_D24_UNORM_PACK32:
        case PixelFormat::D16_UNORM_S8_UINT:
        case PixelFormat::D24_UNORM_S8_UINT:
        case PixelFormat::D32_SFLOAT_S8_UINT:
            return ViewAspect::Depth | ViewAspect::Stencil;
        case PixelFormat::S8_UINT:
            return ViewAspect::Stencil;
        default:
            return ViewAspect::Color;
        }
    }

    GPUTextureViewType default_texture_view_type(GPUTextureType type)
    {
        switch (type)
        {
        case GPUTextureType::Tex1D:
            return GPUTextureViewType::Tex1D;
        case GPUTextureType::Tex2D:
            return GPUTextureViewType::Tex2D;
        case GPUTextureType::Tex3D:
            return GPUTextureViewType::Tex3D;
        default:
            AssertUnreachable();
            return GPUTextureViewType::Tex1D;
        }
    }
} // namespace

GPUTextureSubresource default_texture_subresource(const GPUTextureProperties& properties)
{
    return GPUTextureSubresource{
        .aspect = default_view_aspect(properties.format),
        .mip_offset = 0,
        .mip_count = properties.mip_count,
        .layer_offset = 0,
        .layer_count = properties.layer_count,
    };
}

GPUTextureView default_texture_view(const GPUTextureProperties& properties)
{
    return GPUTextureView{
        .type = default_texture_view_type(properties.type),
        .format = properties.format,
        .subresource = default_texture_subresource(properties),
    };
}
} // namespace Reaper

File: src/renderer/texture/gpu_texture_view.zig
// Port of src/renderer/texture/GPUTextureView.{h,cpp}

const vk = @import("vulkan");

const GPUTextureProperties = @import("gpu_texture_properties.zig").GPUTextureProperties;

pub const ViewAspect = packed struct(u32) {
    color: bool = false,
    depth: bool = false,
    stencil: bool = false,
    _reserved: u29 = 0,

    pub const color_only: ViewAspect = .{ .color = true };
    pub const depth_only: ViewAspect = .{ .depth = true };
    pub const depth_stencil: ViewAspect = .{ .depth = true, .stencil = true };
    pub const stencil_only: ViewAspect = .{ .stencil = true };

    pub fn toVk(self: ViewAspect) vk.ImageAspectFlags {
        return .{
            .color_bit = self.color,
            .depth_bit = self.depth,
            .stencil_bit = self.stencil,
        };
    }
};

pub const GPUTextureSubresource = struct {
    aspect: ViewAspect = .color_only,
    mip_offset: u32 = 0,
    mip_count: u32 = 1,
    layer_offset: u32 = 0,
    layer_count: u32 = 1,
};

pub const GPUTextureViewType = enum {
    tex_1d,
    tex_2d,
    tex_3d,
    tex_cube,
    tex_1d_array,
    tex_2d_array,
    tex_cube_array,
};

pub const GPUTextureView = struct {
    type: GPUTextureViewType = .tex_2d,
    format: vk.Format = .undefined,
    subresource: GPUTextureSubresource = .{},
};

pub fn defaultTextureSubresourceOneColorMip(mip_index: u32, layer_index: u32) GPUTextureSubresource {
    return .{
        .aspect = .color_only,
        .mip_offset = mip_index,
        .mip_count = 1,
        .layer_offset = layer_index,
        .layer_count = 1,
    };
}

pub fn defaultViewAspect(format: vk.Format) ViewAspect {
    return switch (format) {
        .d16_unorm, .d32_sfloat => .depth_only,
        .x8_d24_unorm_pack32, .d16_unorm_s8_uint, .d24_unorm_s8_uint, .d32_sfloat_s8_uint => .depth_stencil,
        .s8_uint => .stencil_only,
        else => .color_only,
    };
}

pub fn defaultTextureSubresource(properties: GPUTextureProperties) GPUTextureSubresource {
    return .{
        .aspect = defaultViewAspect(properties.format),
        .mip_offset = 0,
        .mip_count = properties.mip_count,
        .layer_offset = 0,
        .layer_count = properties.layer_count,
    };
}

pub fn defaultTextureView(properties: GPUTextureProperties) GPUTextureView {
    return .{
        .type = switch (properties.type) {
            .tex_1d => .tex_1d,
            .tex_2d => .tex_2d,
            .tex_3d => .tex_3d,
        },
        .format = properties.format,
        .subresource = defaultTextureSubresource(properties),
    };
}
